fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,263 @@
-            if len(params) == 0:
-                continue
-
-            for p in params:
-                if p.grad is None:
-                    p.grad = torch.zeros_like(p)
-
-            global_rank = OSS.get_global_rank(group, rank)
-
-            # Copy small gradients into per-GPU buffers and then async reduce
-            i_bucketed = 0  # the number of tensors packed in the buffer
-            offset = 0
-
-            # Since all the parameters are already sorted per increasing size, we only need to consider the first ones.
-            while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
-                end = offset + params[i_bucketed].numel()
-                buffer[offset:end].copy_(params[i_bucketed].grad.data.view(-1))  # type: ignore
-                offset = end
-                i_bucketed += 1
-
-            if i_bucketed > 0:
-                buffer.div_(world_size)
-                bucket_requests.append(
-                    (
-                        dist.reduce(tensor=buffer, dst=global_rank, group=group, async_op=True),  # type: ignore
-                        rank,
-                    )
+                # Make sure that this is not fired twice
+                self._grad_to_be_reduced[index] = False
+
+                # Copy to the flat buffer, update the buffer state
+                bucket = optimizer.buckets[param.device][dst_rank]
+
+                assert bucket.append(param, use_gradient=True), "%s - %s - %s" % (
+                    bucket.max_size,
+                    bucket.current_offset,
+                    param.grad.numel(),
                 )
 
-            # Directly reduce the other grads
-            for p in params[i_bucketed:]:
-                p.grad = cast(Tensor, p.grad)
-                if p.grad.requires_grad:
-                    raise RuntimeError("DistributedDataParallel only works with gradients that don't require grad")
+                if bucket.full():
 
-                p.grad.div_(world_size)
-                requests.append(dist.reduce(tensor=p.grad, dst=global_rank, group=group, async_op=True))  # type: ignore
+                    def unwrap() -> None:
+                        for flat in bucket.params:
+                            if dst_rank != self.global_rank:
+                                # this rank is not the owner, release the grad
+                                flat.param.grad = None
+                            else:
+                                # this rank is the owner, unroll the results
+                                assert flat.param.grad is not None
 
-        # Unroll the initial packed small gradients, as soon as possible
-        for future, rank in bucket_requests:
-            future.wait()
+                                flat.param.grad.data.copy_(
+                                    bucket.buffer[flat.start : flat.stop].view_as(flat.param.data)
```

Comment:
maybe using `non_blocking=True` will help a bit here.

Reply:
ah, I checked that in the doc in the past and it was mentioning that it was useful for cpu<>gpu, but I can test this :) thanks !